qiskit/chemistry/ground_state_calculation/mes_factory.py

```diff
@@ -0,0 +1,60 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The minimum eigensolver factory for ground state calculation algorithms."""
+
+from qiskit.aqua import QuantumInstance
+from qiskit.aqua.algorithms import MinimumEigensolver, VQE
+from qiskit.chemistry.components.variational_forms import UCCSD
+from qiskit.chemistry.qubit_transformations import QubitOperatorTransformation
+from qiskit.chemistry.components.initial_states import HartreeFock
+
+
+class MESFactory:
+    """A factory to construct a minimum eigensolver suitable for a qubit operator transformation.
+    """
+
+    def __init__(self, quantum_instance: QuantumInstance) -> None:
+        """
+        Args:
+            quantum_instance: The quantum instance used in the minimum eigensolver.
+        """
+        self._quantum_instance = quantum_instance
+
+    def get_solver(self, transformation: QubitOperatorTransformation) -> MinimumEigensolver:
+        """Returns a minimum eigensolver, based on the qubit operator transformation.
+
+        By default the VQE with a UCCSD wavefunction ansatz is returned.
+
+        Args:
+            transformation: The qubit operator transformation.
+
+        Returns:
+            A minimum eigensolver suitable to compute the ground state of the molecule transformed
+            by ``transformation``.
+        """
+
+        num_orbitals = transformation._molecule_info['num_orbitals']
```

Comment:
I think the advantage of a factory is that it allows us to type check. E.g. AdaptVQE requires UCCSD as ansatz and VQE as eigensolver -- with a factory we can enforce this. @adekusar-drl I think you were also a proponent of the factory, right?

As for the overhead of a callback vs. factory: it's a little easier to write a function instead of a class but the class here will only be 2 more lines of code, which is probably ok (1) given the fact that we can check types and (2) this is likely a not much used feature. And even if you use, the overhead would come from learning how to use the callback/factory, not from coding it 😉

Reply:
There are options in VQE and UCCSD that a user might want to set, simple thing like optimizer or which expectation object. If the factory here was for VQE+UCCSD where a number of the params were exposed so things like this were settable then a factory would be less of a concern for me. We could have one for HF+particle preserving ansatz too.
Without such parameterization having to wrap your custom instantiation of VQE+UCCSD as a factory seems likes its just plain clunky. Its the one method that todate has been done as a callback. As some factory that you have some degrees of flexibility what it builds that seems more worth it. The argument against was is VQE or UCCSD change then the factory might have to change too. Sure but it does not like too much of a downside. Parameters might be added and the factory would not expose them until they were added, sure if parameters were deprecated it would have to change. But these are algos we expose to user so we evidently want to keep their API reasonably stable.
And lets look at enforcement. Currently VQEAdapt checks whether the factory has VQE plus UCCSD. It cannot do a strict type check to VQEUSSCDFactory since that is just one config of VQE+UCSSD and we are going to require the user to build their own out to configure VQE etc in way they want.

Reply:
We now expose some of the VQE arguments through the Factory.